### PR TITLE
Fix skeletal animation sample() does not work as intended

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: cocos-creator/3d-tasks#

Changelog:
 * This PR fixes:
   - The first frame of skeletal animation is not played.

   - If `SkeletalAnimationState.play()` is not called, the `SkeletalAnimationState.sample()` takes no effect. (https://forum.cocos.org/t/topic/127457/198?u=shrinktofit)

 * Also optimized:
 
  - The mesh renderer should be child of skeletal animation.

### CODE REVIEW

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 1]
```diff
@@ -83,19 +83,13 @@ export class SkeletalAnimationState extends AnimationState {
         this._frames = frames - 1;
         this._animInfo = this._animInfoMgr.getData(root.uuid);
         this._bakedDuration = this._frames / samples; // last key
+        this.setUseBaked(baked);
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts:86]: This ensure the bake option(`true` vs `false`) of animation state is up-to-date with skinned mesh renderer, whenever.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 2]
```diff
@@ -83,19 +83,13 @@ export class SkeletalAnimationState extends AnimationState {
...
     }
 
-    public onPlay () {
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts:88]: Before this PR: do "sync bake option" on playing.

This PR: Sync at `initialize()`, sync bake option each time it changes.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 3]
```diff
@@ -83,19 +83,13 @@ export class SkeletalAnimationState extends AnimationState {
...
-        super.onPlay();
-        const baked = this._parent!.useBakedAnimation;
-        if (baked) {
+    public setUseBaked (useBaked: boolean) {
+        if (useBaked) {
             this._sampleCurves = this._sampleCurvesBaked;
             this.duration = this._bakedDuration;
-            this._animInfoMgr.switchClip(this._animInfo!, this.clip);
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts:94]: Moved to below.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 4]
```diff
@@ -168,8 +162,24 @@ export class SkeletalAnimationState extends AnimationState {
     private _sampleCurvesBaked (time: number) {
         const ratio = time / this.duration;
         const info = this._animInfo!;
+
+        // Ensure I'm the one on which the anim info is sampling.
+        if (!this._animInfoMgr.isCurrentlySampling(info, this.clip)) {
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts]: @YunHsiao Please review the main fix here. I'll do the check-switch each time the animation is sampled.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 5]
```diff
@@ -168,8 +162,24 @@ export class SkeletalAnimationState extends AnimationState {
...
+            this._animInfoMgr.switchClip(this._animInfo!, this.clip);
+
+            const users = this._parent!.getUsers();
+            users.forEach((user) => {
+                user.uploadAnimation(this.clip);
+            });
+        }
+
         const curFrame = (ratio * this._frames + 0.5) | 0;
-        if (curFrame === info.data[0]) { return; }
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts:172]: @YunHsiao Please review the fix here. This cause the first frame always be skipped.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts [hunk 6]
```diff
@@ -168,8 +162,24 @@ export class SkeletalAnimationState extends AnimationState {
...
+        if (info.hasSampled && curFrame === info.data[0]) {
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-state.ts]: And the fix.

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts [hunk 7]
```diff
@@ -457,6 +457,11 @@ export interface IAnimInfo {
     buffer: Buffer;
     data: Float32Array;
     dirty: boolean;
+    currentClipRef: WeakSet<AnimationClip>;
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: I want to keep a record of `AnimationClip`, something like `WeakRef`. If you have better opinion, please.

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts [hunk 8]
```diff
@@ -457,6 +457,11 @@ export interface IAnimInfo {
...
+    /**
+     * Has sampled at least once since last switching.
+     */
+    hasSampled: boolean;
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: Fix about the first frame BUG.

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts [hunk 9]
```diff
@@ -506,7 +511,16 @@ export class JointAnimationInfo {
         }
     }
 
+    public isCurrentlySampling (info: IAnimInfo, clip: AnimationClip) {
+        return info.currentClipRef.has(clip);
+    }
+
     public switchClip (info: IAnimInfo, clip: AnimationClip | null) {
+        info.hasSampled = false;
+        info.currentClipRef = new WeakSet();
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: `WeakSet` doesn not provide a `clear()`

File: cocos/3d/skinned-mesh-renderer/skinned-mesh-renderer.ts [hunk 10]
```diff
@@ -146,43 +154,31 @@ export class SkinnedMeshRenderer extends MeshRenderer {
...
     }
 
-    private _animation: SkeletalAnimation | null = null;
-
     private _tryBindAnimation () {
         const { _skinningRoot: skinningRoot } = this;
         if (!skinningRoot) {
             return;
         }
+
```
Comment (shrinktofit) [inline cocos/3d/skinned-mesh-renderer/skinned-mesh-renderer.ts:167]: Corresponding to https://github.com/cocos-creator/engine/pull/10006/files#diff-6b7762c60bea81084e63372b54ddae7766cb7e91ecc85de1977d3a51664bcd25R170-R181 . They should take same effect.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 6 above)
Comment (YunHsiao) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: I think initializing the anim data to a special illegal value would do, saves the ad-hoc flag test

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts [hunk 11]
```diff
@@ -506,7 +511,16 @@ export class JointAnimationInfo {
         }
     }
 
+    public isCurrentlySampling (info: IAnimInfo, clip: AnimationClip) {
```
Comment (pandamicro) [inline cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: `isSampling` is enough, the present continuous tense is the same as current

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts (hunk 9 above)
Comment (pandamicro) [reply cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: It has been created previously, why replace it here

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts (hunk 7 above)
Comment (YunHsiao) [reply cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: Just store the clip itself. The whole anim info object will be GCed once the component is destroyed.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 2 above)
Comment (YunHsiao) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts:88]: Are you sure this works for both baked & sampled animations? I think the sampled branch has some init code that still needs to be executed on each play event.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 4 above)
Comment (pandamicro) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: This looks very strange, previously `onPlay` is only invoked in the beginning, but `_sampleCurvesBaked` is invoked in every animation frame, it just doesn't seem elegant enough

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 6 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: > 

The only illegal is `NaN`, could I?

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts (hunk 7 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: Got it .

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts (hunk 9 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: Fixed.

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts (hunk 11 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-utils.ts]: Fixed.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 2 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts:88]: > Are you sure this works for both baked & sampled animations? I think the sampled branch has some init code that still needs to be executed on each play event.

The init code of baked branch was moved to `setUseBaked`, which is invoked at `SkeletalAnimationState.initialize()` and `set SkeletalAnimation.useBakedAnimation()`. Is there something I missed?

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 4 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: There are two methods that trigger the sampling process:

- `SkeletalAnimationState.play()`/`SkeletalAnimationState.resume()` - Called once, do sampling every frame

- `SkeletalAnimationState.sample()` - Called once, do once sampling immediately.

So I captured the sampling process.

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 6 above)
Comment (YunHsiao) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: I think any negative number would do?

**Review by pandamicro — APPROVED** [review]

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 2 above)
Comment (YunHsiao) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts:88]: I mean what about the `else` branch, under real-time sampled animation mode?

File: cocos/3d/skeletal-animation/skeletal-animation-utils.ts [hunk 12]
```diff
@@ -507,7 +508,8 @@ export class JointAnimationInfo {
     }
 
     public switchClip (info: IAnimInfo, clip: AnimationClip | null) {
-        info.data[0] = 0;
+        info.currentClip = clip;
+        info.data[0] = -1;
```
Comment (shrinktofit) [inline cocos/3d/skeletal-animation/skeletal-animation-utils.ts:512]: @YunHsiao

File: cocos/3d/skeletal-animation/skeletal-animation-state.ts (hunk 6 above)
Comment (shrinktofit) [reply cocos/3d/skeletal-animation/skeletal-animation-state.ts]: OK

**Review by YunHsiao — APPROVED** [review]

**Pull request merged**